File: 206. Reverse Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    // iterative: O(N) time, O(1) space, 45.71%, ok
    ListNode* reverseList(ListNode* head) {
        if(!head) return NULL;
        if(!head->next) return head;  // bug: only one element, directly return.
        ListNode* next = head->next;
        ListNode* pre = NULL;
        head->next = pre;
        pre = head;
        head = next;
        while(head->next){
            next = head->next;
            head->next = pre;
            pre = head;
            head = next;
        }
        head->next = pre;
        return head;
    }

    // iterative with more elegancy: 45.71%, ok
    ListNode* reverseList(ListNode* head) {
        if(!head) return NULL;
        ListNode* prev = NULL;
        while(head && head->next){
            ListNode* temp = head->next;
            head->next = prev;
            prev = head;
            head = temp;
        }
        head->next = prev;
        return head;
    }

    // (STUCK) recursive: 45.71%
    // O(N) time, key is to work backwards
    ListNode* reverseList(ListNode* head) {
        if(!head || !head->next) return head;
        ListNode* nextNode = reverseList(head->next);
        head->next->next = head;
        head->next = NULL;
        return nextNode;
    }

};

File: 206. Reverse Linked List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    
    # Iteratively: O(N) time, O(1) space
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head or not head.next: return head
        prev = None
        while head and head.next:
            temp = head.next
            head.next = prev
            prev = head
            head = temp
        head.next = prev
        return head

    # Recursively: O(N) time, O(1) space
    def reverseList(self, head):
        if not head or not head.next:
            return head
        node = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        return node